### ISSUE TITLE
SB16 command 0xf9 doesn't have anything to do with the ASP

### ISSUE THREAD

**Issue by cracyc** [joncampbell123/dosbox-x#1044]
SB16 command 0xf9 returns the contents of the 8051 ram at the address of the second byte and 0xfa sets value at the address of the second byte to the third byte.

**Comment by joncampbell123** [comment]
Got it. I will release DOSBox-X this Saturday and then revise the sblaster.cpp source code.

I will also look into what exactly the Win98 SB16 driver expects out of cards that do not implement ASP.

**Comment by cracyc** [comment]
Here's a partial memory map if you want to include docs for the addresses that the drivers use:

```
RAM:000B .equ play_mode,0xB             
RAM:000C .equ dma16_len_lo,0xC          
RAM:000D .equ dma16_len_hi,0xD          
RAM:000E .equ dma8_len_lo,0xE           
RAM:000F .equ dma8_len_hi,0xF           
RAM:0011 .equ single_cycle_len_lo,0x11  
RAM:0012 .equ single_cycle_len_hi,0x12  
RAM:0013 .equ rate_low,0x13             
RAM:0014 .equ rate_high,0x14            
RAM:0020 .equ cur_cmd,0x20              
RAM:0021 .equ len_lo,0x21               
RAM:0022 .equ len_hi,0x22               
RAM:0023 .equ dsp_stat,0x23             
RAM:0024 .equ dsp_aux_stat,0x24         
RAM:0025 .equ prot_reg0,0x25            
RAM:0026 .equ prot_reg1,0x26            
RAM:002A .equ test_reg,0x2A             
RAM:002C .equ play_mode_8,0x2C          
RAM:002D .equ play_mode_16,0x2D         
RAM:0030 .equ last_cmd,0x30             
RAM:0037 .equ cur_rate,0x37             
RAM:003A .equ stack,0x3A
```

**Comment by joncampbell123** [comment]
@cracyc Nice! Where did you find that documentation?

**Comment by cracyc** [comment]
My own disassembly from when I was emulating it low level in mame.  Here it is if you want to look at it.
[sb16asm.zip](https://github.com/joncampbell123/dosbox-x/files/3045124/sb16asm.zip)

**Comment by joncampbell123** [comment]
I have some working DOS machines with various SB16 cards, perhaps I can poke at them with DOSLIB and see what the values are.

One interesting quirk of SB16 I could verify with this is that SB16 cards reduce the precision of the sample rate when it actually uses it.

For example, this graph generated using a program in DOSLIB to test all 65536 values of the rate command, rendered with GNUplot.

http://hackipedia.org/raw/Computer/Platform/PC%2c%20IBM%20compatible/Sound%20and%20Music/Creative%20Labs/Sound%20Blaster/Collection/DOSLIB%20documentation/Creative%20Sound%20Blaster%2016%20ViBRAC%20PnP%20CT4180/graphs/rates/Creative%20Sound%20Blaster%2016%20ViBRAC%20PnP%20CT4180%2etxt%2esb16%2epng

**Comment by joncampbell123** [comment]
And here's a pre-PnP version of the SB16 when the DSP apparently had bugs in the sample rate handling. Note the periodic dips.

http://hackipedia.org/raw/Computer/Platform/PC%2c%20IBM%20compatible/Sound%20and%20Music/Creative%20Labs/Sound%20Blaster/Collection/DOSLIB%20documentation/Creative%20Sound%20Blaster%2016%20CT1740/graphs/rates/Creative%20Sound%20Blaster%2016%20CT1740%2etxt%2esb16%2epng

However this model also has a working ASP chip on it.

**Comment by cracyc** [comment]
The rate PLL has only 8bits of resolution so the stair step is easy to explain.  The formula for converting the PLL rate to sample rate is 46615120/1024/256 * pllrate and the conversion in the rom is rate * 23 / 4096.  There'll then be error as the first is 177.8 and the second is 178 but that doesn't explain the dips (a bug in the PLL?).

**Comment by joncampbell123** [comment]
My guess is that it's a bug in the ROM. A bug in the PLL would show dips that start and stop precisely on the "steps"

**Comment by cracyc** [comment]
The other thing of note is the 0xe2 command which is much too complex in dosbox.  The real algorithm is just:

```
a = 0xaa
b = 0x69

a += read_byte() ^ b
b = ror(b, 2)
return a
```

**Comment by joncampbell123** [comment]
> The other thing of note is the 0xe2 command which is much too complex in dosbox. The real algorithm is just:
> 
> ```
> a = 0xaa
> b = 0x69
> 
> a += read_byte() ^ b
> b = ror(b, 2)
> return a
> ```

I'm going to have to postpone implementing that because my initial implementation of that seems to produce results inconsistent with the current DOSBox-X implementation.

These results were made by writing out(0x22C,0xe2), out(0x22C,0x00) repeatedly after DSP reset.

First one is the DOSBox-X current implementation, second one is code based on your pseudocode.

LOG: SB E2 0x40 0x13
LOG: SB E2 0xe5 0x6d
LOG: SB E2 0x4e 0x03
LOG: SB E2 0xa8 0xa8
LOG: SB E2 0x3e 0x11

Perhaps my next DOSLIB project should be a program to run this command repeatedly to help validate what it is on real SB16 hardware.

**Comment by joncampbell123** [comment]
Bit8u valadd;
        Bit8u valxor;

    sb.e2.valadd=0xaa;
    sb.e2.valxor=0x69;

             sb.e2.valadd += sb.dsp.in.data[0] ^ sb.e2.valxor;
             sb.e2.valxor = (sb.e2.valxor >> 2u) | (sb.e2.valxor << 6u);

**Comment by joncampbell123** [comment]
I just finished a program to show the sequence of bytes coming back from command 0xE2. I'll pull out the old DOS machine and see how a SB16 responds. Here is the result from within DOSBox-X's current emulation.

https://github.com/joncampbell123/doslib/blob/master/hw/sndsb/e2seq.c

![command_000](https://user-images.githubusercontent.com/6245486/55673133-7718ef80-5858-11e9-8547-dfaf0cbde8b9.png)

**Comment by joncampbell123** [comment]
Keep in mind DOSBox-X's initial state is count = 0 value = 0xAA for command E2 emulation.

**Comment by joncampbell123** [comment]
Hang on a sec.

Looking at E2_incr_table in sblaster.cpp the sign of the value added does correspond to 0x69, however remember that those values are access from LSB to MSB of the DSP data byte written after 0xE2.

I get inconsistent results if I follow your pseudocode with b = 0x69, however I get consistent results if I reverse the bits and use 0x96.

**Comment by cracyc** [comment]
test-sbp.exe which is the only program I have that uses doesn't reset the card yet seems to work no matter what state it's in. Since it'll cycle though 0x96,0xa5,0x69,0x5a it may not matter which is first (BTW the ati stereo f/x sbpro clone uses that table in that order for it's cmd 0xe2).  On a real card you can use cmd 0xf9 to read the states from bytes 0x25 and 0x26.

**Comment by joncampbell123** [comment]
It checks out on real hardware.

![vlcsnap-2019-04-06-14h23m53s024](https://user-images.githubusercontent.com/6245486/55675477-ee11b080-5877-11e9-9515-a4f53478e83f.png)

I'll add code to the SNDSB TEST program to start playing with commands 0xE9 and 0xFA.

It will have the same function as the code already there to play with the extra hardware registers on ESS688 sound cards.

**Comment by joncampbell123** [comment]
So here's the test results with the updated SNDSB TEST program and SB16 8051 memory.

There's all sorts of random bytes in there. You can clear most of them while playing DMA audio with no ill effects. Some of the bytes change during playback (but very few). However if you write dsp aux status (0x24) with bit 4 zero (16-bit playback) or bit 2 zero (8-bit playback) the DSP stops at the next interrupt. If you play 16-bit DMA and write 0x11 instead of 0x10, playback stops and the card just cycles DMA and playback endlessly with no sound.

Sound Blaster AWE64 ISA PnP.

![vlcsnap-2019-04-06-15h01m54s568](https://user-images.githubusercontent.com/6245486/55675819-1ea81900-587d-11e9-985f-4aecb4871a80.png)

**Comment by joncampbell123** [comment]
Here's the 8051 RAM after starting TEST.EXE, but before starting playback.
![vlcsnap-2019-04-06-15h06m06s891](https://user-images.githubusercontent.com/6245486/55675844-8fe7cc00-587d-11e9-9e6e-3015541847f0.png)

**Comment by joncampbell123** [comment]
Actually it seems playback did not fill the RAM with random values.

Perhaps those were there at boot-up?

On the ASP card the ASP RAM is also filled with random values after cold boot as well.

**Comment by joncampbell123** [comment]
I tried writing values to the "sample rate" field while playing. It didn't have any effect. However that field changes when the sample rate is given through DSP commands.

**Comment by cracyc** [comment]
That value is written to the asic when playback starts so you'd probably have to stop and start playing for it to take effect.

**Comment by joncampbell123** [comment]
Quick question, if you know: Are commands 0xF9/0xFA specific to the SB16 or can they be used on the SB Pro and SB 2.0 cards as well?

If you don't know, that's fine. I can poke at those cards later on to see if those commands exist.

**Comment by cracyc** [comment]
I don't know but suspect not.

**Comment by cracyc** [comment]
RE https://github.com/joncampbell123/dosbox-x/commit/97843fd3e92515727786c857d1f9f6570fa04f70, the time constant value isn't saved in ram.  It just converts the tc rate directly to the pll rate from a lookup table.

```
.byte 0x15, 0x16, 0x16, 0x16, 0x16, 0x16, 0x16, 0x16, 0x16; 0
.byte 0x16, 0x16, 0x16, 0x17, 0x17, 0x17, 0x17, 0x17, 0x17; 9
.byte 0x17, 0x17, 0x17, 0x17, 0x18, 0x18, 0x18, 0x18, 0x18; 18
.byte 0x18, 0x18, 0x18, 0x18, 0x18, 0x19, 0x19, 0x19, 0x19; 27
.byte 0x19, 0x19, 0x19, 0x19, 0x1A, 0x1A, 0x1A, 0x1A, 0x1A; 36
.byte 0x1A, 0x1A, 0x1A, 0x1B, 0x1B, 0x1B, 0x1B, 0x1B, 0x1B; 45
.byte 0x1B, 0x1B, 0x1C, 0x1C, 0x1C, 0x1C, 0x1C, 0x1C, 0x1C; 54
.byte 0x1D, 0x1D, 0x1D, 0x1D, 0x1D, 0x1D, 0x1E, 0x1E, 0x1E; 63
.byte 0x1E, 0x1E, 0x1E, 0x1F, 0x1F, 0x1F, 0x1F, 0x1F, 0x1F; 72
.byte 0x20, 0x20, 0x20, 0x20, 0x20, 0x21, 0x21, 0x21, 0x21; 81
.byte 0x21, 0x22, 0x22, 0x22, 0x22, 0x22, 0x23, 0x23, 0x23; 90
.byte 0x23, 0x24, 0x24, 0x24, 0x24, 0x25, 0x25, 0x25, 0x25; 99
.byte 0x26, 0x26, 0x26, 0x26, 0x27, 0x27, 0x27, 0x27, 0x28; 108
.byte 0x28, 0x28, 0x29, 0x29, 0x29, 0x29, 0x2A, 0x2A, 0x2A; 117
.byte 0x2B, 0x2B, 0x2B, 0x2C, 0x2C, 0x2D, 0x2D, 0x2D, 0x2E; 126
.byte 0x2E, 0x2E, 0x2F, 0x2F, 0x30, 0x30, 0x30, 0x31, 0x31; 135
.byte 0x32, 0x32, 0x33, 0x33, 0x34, 0x34, 0x35, 0x35, 0x36; 144
.byte 0x36, 0x37, 0x37, 0x38, 0x38, 0x39, 0x39, 0x3A, 0x3B; 153
.byte 0x3B, 0x3C, 0x3D, 0x3D, 0x3E, 0x3F, 0x3F, 0x40, 0x41; 162
.byte 0x42, 0x42, 0x43, 0x44, 0x45, 0x46, 0x47, 0x48, 0x49; 171
.byte 0x49, 0x4A, 0x4B, 0x4D, 0x4E, 0x4F, 0x50, 0x51, 0x52; 180
.byte 0x53, 0x55, 0x56, 0x57, 0x59, 0x5A, 0x5C, 0x5D, 0x5F; 189
.byte 0x60, 0x62, 0x64, 0x66, 0x68, 0x6A, 0x6C, 0x6E, 0x70; 198
.byte 0x72, 0x75, 0x77, 0x7A, 0x7C, 0x7F, 0x82, 0x85, 0x89; 207
.byte 0x8C, 0x90, 0x93, 0x97, 0x9C, 0xA0, 0xA5, 0xAA, 0xAF; 216
.byte 0xB5, 0xBB, 0xC1, 0xC8, 0xD0, 0xD8, 0xE0, 0xEA, 0xF4; 225
.byte 0xFF, 0xFF        ; 234

```

**Comment by joncampbell123** [comment]
Done. Thank you.

Which part of the card enforces the minimum (5000Hz) and maximum (45454Hz) sample rates?

I also just noticed from the graphs that SB and SB Pro have a minimum of 4000Hz, but the SB16 has a minimum of 5000Hz IF you set the sample rate using DSP command 0x41.

Regarding the SB16 with the "dips" in the sample rate chart, another chart shows a spike in sample rate at one specific time constant on the same card.

http://hackipedia.org/raw/Computer/Platform/PC%2c%20IBM%20compatible/Sound%20and%20Music/Creative%20Labs/Sound%20Blaster/Collection/DOSLIB%20documentation/Creative%20Sound%20Blaster%2016%20CT1740/graphs/rates/Creative%20Sound%20Blaster%2016%20CT1740%2etxt%2esb1%2epng

**Comment by joncampbell123** [comment]
More here if interested, test results using DOSLIB tools.

http://hackipedia.org/browse.cgi/Computer/Platform/PC%2c%20IBM%20compatible/Sound%20and%20Music/Creative%20Labs/Sound%20Blaster/Collection/DOSLIB%20documentation/

**Comment by cracyc** [comment]
> Regarding the SB16 with the "dips" in the sample rate chart, another chart shows a spike in sample rate at one specific time constant on the same card.

If you're right that it's a bug in the rom then it must be specific to that rom version (the rom from mame is from a ct1741) we'd probably have to have a dump from a ct1740 to know for sure.  I've examined the rom for some way to dump it in software but there's nothing that's obvious although I'm far from an expert on that kind of thing.

>Which part of the card enforces the minimum (5000Hz) and maximum (45454Hz) sample rates?

It compares the high byte of the rate to 0x13 and 0xb1 and sets the pll to 0x1c if below or 0xff if above.

**Comment by joncampbell123** [comment]
While we're talking about SB16, you might like this issue. I have some notes I've taken already posted in it.

https://github.com/joncampbell123/dosbox-x/issues/1047

**Comment by cracyc** [comment]
It might be interesting to set the rate to the outlier values and check address 0x37 in the ram to see if it's clearly invalid.

**Comment by joncampbell123** [comment]
@cracyc Exactly: https://github.com/joncampbell123/doslib/blob/master/hw/sndsb/ts_sb16a.c

**Comment by joncampbell123** [comment]
@cracyc Fun discovery tonight with DOSLIB testing: The PC-98 version of the Sound Blaster 16 card has DSP v4.5 and it has the same exact bug as the IBM PC/AT version I noted with sample rates.

https://github.com/joncampbell123/dosbox-x/issues/1047#issuecomment-492930276

**Comment by cracyc** [comment]
> If you don't know, that's fine. I can poke at those cards later on to see if those commands exist.

Now that there's a dump of a Sound Blaster 2.0, 0xf8 does something different and 0xf9 doesn't exist (in fact it's incompletely decoded so if bit 3 is set it does 0xf8, bit 2 0xf4 etc. in decending order).

**Comment by joncampbell123** [comment]
@cracyc I saw that, Tube Time on Twitter. Is the dump available for download?

**Comment by joncampbell123** [comment]
https://www.vogons.org/viewtopic.php?f=62&t=61098&start=280
